### PR TITLE
[Calyx] Add helper functions to access MemoryOp ports (#1635)

### PR DESCRIPTION
**Pull request by mortbopet**
Add helper functions to access MemoryOp ports (#1635).

### CODE REVIEW

File: include/circt/Dialect/Calyx/CalyxPrimitives.td [hunk 1]
```diff
@@ -99,4 +99,15 @@ def MemoryOp : CalyxPrimitive<"memory", [
       "ArrayRef<int64_t>":$addrSizes
     )>
   ];
+
+  let extraClassDeclaration = [{
+    size_t getNumAddresses() { return addrSizes().size(); }
```
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/CalyxPrimitives.td:104]: tiny nit: `getNumDimensions` ? I guess yours would be more in line with the op's names :)

File: include/circt/Dialect/Calyx/CalyxPrimitives.td [hunk 2]
```diff
@@ -99,4 +99,15 @@ def MemoryOp : CalyxPrimitive<"memory", [
...
+
+    ValueRange addrPorts() { return getResults().take_front(getNumAddresses()); }
+    Value addr(size_t i) { assert(i<getNumAddresses()); return getResults()[i]; }
```
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/CalyxPrimitives.td:107]: Add a message to assert `... && "Out of bounds of address ports for this memory."`

File: include/circt/Dialect/Calyx/CalyxPrimitives.td [hunk 3]
```diff
@@ -99,4 +99,15 @@ def MemoryOp : CalyxPrimitive<"memory", [
       "ArrayRef<int64_t>":$addrSizes
     )>
   ];
+
+  let extraClassDeclaration = [{
+    size_t getNumAddresses() { return addrSizes().size(); }
+
+    ValueRange addrPorts() { return getResults().take_front(getNumAddresses()); }
```
Comment (cgyurgyik) [inline include/circt/Dialect/Calyx/CalyxPrimitives.td:106]: Why not index accessor for this as well?

**Review by cgyurgyik — APPROVED** [review]
A few nits but LGTM. Have you tested these? We don't have any C++ code that uses memory yet, but you can just build one in a random pass and use `llvm::errs() << ...` to verify you're getting the output you want.

Makes me wonder why unit tests don't exist for things like this.

File: include/circt/Dialect/Calyx/CalyxPrimitives.td (hunk 3 above)
Comment (mortbopet) [reply include/circt/Dialect/Calyx/CalyxPrimitives.td:106]: By this, do you mean a `Value addrPort(size_t i)` function?

Comment (cgyurgyik) [reply include/circt/Dialect/Calyx/CalyxPrimitives.td]: Yeah, as you did for `addr(size_t i)`. An accessor by index and a ValueRange accessor for both makes sense to me. When working on frontend stuff for Calyx, I recall using both of these.

**Review by mikeurbach — COMMENTED** [review]
LGTM, thanks for adding this!

**Pull request merged**